[torch] Fix hashing for SymInt

khabinov opened the pull request:
Differential Revision: D53335479

suo commented:
I said internally: I think this is fine. The only required property of hash is that if a==b then hash(a) must equal hash(b).

This meets those requirements, but not sure if there are any other considerations. @ezyang any thoughts?

khabinov committed:
[torch] Fix hashing for SymInt (#119244)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/119244

Differential Revision: D53335479

ezyang reviewed (changes requested):
stop the presses

ezyang commented:
You can't do this. If you hash a SymInt in this way, you can depend on a specific hint of an expression in an observable way without installing an appropriate guard. For example, let's say you use a hash to bucket values, you would silently burn in the bucketing from the particular hints of the SymInt at the time of tracing.

The safe thing to do is force specialization when hashing occurs. This version of the patch is not too difficult to do and I would greatly prefer it.

khabinov commented:
> The safe thing to do is force specialization when hashing occurs.

@ezyang sorry, I think I am a bit lost. Could you please elaborate on what kind of specialization are we talking about here? For SymInt or for SymInt hint or something else?

ezyang commented:
To "specialize" a SymInt means to convert it back into a plain integer. This is done by installing a guard requiring the symbolic quantity in question to be exactly some integer for the graph to be valid.

@suo what does 👀 mean lol

suo commented:
haha, just that it was more complicated than my naive/superficial evaluation thought it would be

khabinov commented:
Will close in favor of #119602
